File: csharp/error-handling/ErrorHandling.cs
using System;

public static class ErrorHandling
{
    public static void HandleErrorByThrowingException() => throw new Exception(nameof(HandleErrorByThrowingException));

    public static int? HandleErrorByReturningNullableType(string input) => int.TryParse(input, out int success) ? success : (int?)null;

    public static bool HandleErrorWithOutParam(string input, out int result) => int.TryParse(input, out result);

    public static void DisposableResourcesAreDisposedWhenExceptionIsThrown(IDisposable disposableObject)
    {
        try
        {
            throw new Exception(nameof(DisposableResourcesAreDisposedWhenExceptionIsThrown));
        }
        finally
        {
            disposableObject.Dispose();
        }
    }
}

File: fsharp/error-handling/ErrorHandling.fs
module ErrorHandling
open System

let handleErrorByThrowingException() = failwith "Error"

let handleErrorByReturningOption (input:string) = 
    try
        int input |> Some
    with
    | _ -> None

let handleErrorByReturningResult input = 
    match handleErrorByReturningOption input with
    | Some n -> Ok n
    | None -> Error "Could not convert input to integer"

let bind switchFunction twoTrackInput = 
    match twoTrackInput with
    | Error s -> Error s
    | Ok s -> switchFunction s

let cleanupDisposablesWhenThrowingException (resource:IDisposable) = 
    try
        failwith "Error"
    finally
        resource.Dispose()